server/base/src/main/java/org/apache/accumulo/server/metadata/RefreshesImpl.java

```diff
@@ -0,0 +1,237 @@
+          tserver);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void add(Collection<RefreshEntry> entries) {
+    Objects.requireNonNull(entries);
+    if (dataLevel == Ample.DataLevel.ROOT) {
+      // expect all of these to be the root tablet, verifying because its not stored
+      Preconditions
+          .checkArgument(entries.stream().allMatch(e -> e.getExtent().equals(RootTable.EXTENT)));
+      Consumer<Map<String,String>> mutator = map -> entries.forEach(refreshEntry -> map
+          .put(refreshEntry.getEcid().canonical(), refreshEntry.getTserver().getHostPortSession()));
+      mutateRootRefreshes(mutator);
+    } else {
+      try (BatchWriter writer = context.createBatchWriter(dataLevel.metaTable())) {
+        for (RefreshEntry entry : entries) {
+          writer.addMutation(createAddMutation(entry));
+        }
+      } catch (MutationsRejectedException | TableNotFoundException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+  }
+
+  @Override
+  public void delete(Collection<RefreshEntry> entries) {
+    Objects.requireNonNull(entries);
+    if (dataLevel == Ample.DataLevel.ROOT) {
+      // expect all of these to be the root tablet, verifying because its not stored
+      Preconditions
+          .checkArgument(entries.stream().allMatch(e -> e.getExtent().equals(RootTable.EXTENT)));
+      Consumer<Map<String,String>> mutator =
+          map -> entries.forEach(refreshEntry -> map.remove(refreshEntry.getEcid().canonical()));
+      mutateRootRefreshes(mutator);
+    } else {
+      try (BatchWriter writer = context.createBatchWriter(dataLevel.metaTable())) {
+        for (RefreshEntry entry : entries) {
+          writer.addMutation(createDeleteMutation(entry));
+        }
+      } catch (MutationsRejectedException | TableNotFoundException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+  }
+
+  @Override
+  public Stream<RefreshEntry> list() {
+    if (dataLevel == Ample.DataLevel.ROOT) {
+      var zooReader = context.getZooReader();
+      byte[] jsonBytes;
+      try {
+        jsonBytes =
+            zooReader.getData(context.getZooKeeperRoot() + RootTable.ZROOT_TABLET_REFRESHES);
+        return fromJson(new String(jsonBytes, UTF_8)).entrySet().stream()
+            .map(e -> new RefreshEntry(ExternalCompactionId.of(e.getKey()), RootTable.EXTENT,
+                new TServerInstance(e.getValue())));
+      } catch (KeeperException | InterruptedException e) {
+        throw new IllegalStateException(e);
+      }
+    } else {
+      Range range = RefreshSection.getRange();
+
+      Scanner scanner;
+      try {
+        scanner = context.createScanner(dataLevel.metaTable(), Authorizations.EMPTY);
+      } catch (TableNotFoundException e) {
+        throw new IllegalStateException(e);
+      }
+      scanner.setRange(range);
+      return scanner.stream().map(this::decode);
+    }
```

Comment:
```suggestion
      try (Scanner scanner = context.createScanner(dataLevel.metaTable(), Authorizations.EMPTY)) {
        scanner.setRange(range);
        return scanner.stream().map(this::decode);
      } catch (TableNotFoundException e) {
        throw new IllegalStateException(e);
      }
    }
```

Reply:
Fixed in dedaa88.  Could not close the scanner in the method as a stream built from the scanner was returned.  So add an onClose call to the stream to close the scanner and then closed the stream in the calling code.